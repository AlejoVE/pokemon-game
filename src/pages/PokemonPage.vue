<template>
	<h1 v-if="!pokemon">Loading...</h1>
	<div v-else>
		<h1>Which pokemon is this?</h1>
		<PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon" />
		<PokemonOptions
			:pokemons="pokemonArr"
			:selected="showPokemon"
			@selection="checkAnswer($event)"
		/>
	</div>
	<template class="fade-in" v-if="showResult">
		<h2>{{ message }}</h2>
		<button @click="newGame">New game</button>
	</template>
</template>

<script>
import PokemonPicture from '@/components/PokemonPicture.vue'
import PokemonOptions from '@/components/PokemonOptions.vue'
import getPokemonsOptions from '@/helpers/getPokemonsOptions.js'

// console.table(getPokemonsOptions())
export default {
	components: {
		PokemonPicture,
		PokemonOptions,
	},
	data() {
		return {
			pokemonArr: [],
			pokemon: null,
			showPokemon: false,
			showResult: false,
			correctAnswer: null,
			message: '',
		}
	},
	methods: {
		async mixPokemonArr() {
			this.pokemonArr = await getPokemonsOptions()
			// x4 so the index is from 0 to 3
			const randonInt = Math.floor(Math.random() * 4)
			this.pokemon = this.pokemonArr[randonInt]
		},
		checkAnswer(pokemonId) {
			this.showPokemon = true
			const isCorrectAnswer = pokemonId === this.pokemon.id
			if (isCorrectAnswer) {
				this.message = 'Correct!'
			} else {
				this.message = `Sorry, it was ${this.pokemon.name}!`
			}
			this.showResult = true
		},
        newGame() {
			this.pokemon = null
			this.showPokemon = false
			this.showResult = false
            this.pokemonArr = []
            this.message = ''
            this.mixPokemonArr()
        }
	},
	computed: {},
	mounted() {
		this.mixPokemonArr()
	},
}
</script>

<style></style>
