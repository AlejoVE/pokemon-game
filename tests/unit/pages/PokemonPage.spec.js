import { shallowMount, mount } from '@vue/test-utils'
import PokemonPage from '@/pages/PokemonPage'
import { pokemons } from '../mocks/pokemons'

describe('PokemonPage Component', () => {
	let wrapper

	beforeEach(() => {
		wrapper = shallowMount(PokemonPage)
	})

	test('should call mixPokemonArr on mount ', () => {
		const mixPokemonArrSpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArr')
		const wrapper = shallowMount(PokemonPage)

		expect(mixPokemonArrSpy).toHaveBeenCalled()
	})

	test('should match snapshot when pokemon load', () => {
		const wrapper = shallowMount(PokemonPage, {
			data() {
				return {
					pokemonArr: pokemons,
					pokemon: pokemons[0],
					showPokemon: false,
					showResult: false,
					correctAnswer: null,
					message: '',
				}
			},
		})

		expect(wrapper.html()).toMatchSnapshot()
	})

	test('should show  PokemonPicture & PokemonOptions components', () => {
		const wrapper = shallowMount(PokemonPage, {
			data() {
				return {
					pokemonArr: pokemons,
					pokemon: pokemons[0],
					showPokemon: false,
					showResult: false,
					correctAnswer: null,
					message: '',
				}
			},
		})

		const PokemonPicture = wrapper.find('pokemon-picture-stub')
		const PokemonOptions = wrapper.find('pokemon-options-stub')
		expect(PokemonPicture.exists()).toBeTruthy()
		expect(PokemonOptions.exists()).toBeTruthy()

		expect(PokemonPicture.attributes('pokemonid')).toBe('1')
		expect(PokemonOptions.attributes('pokemons')).toBeTruthy()
	})

	test('Tests in checkAnswer', async () => {
		const wrapper = shallowMount(PokemonPage, {
			data() {
				return {
					pokemonArr: pokemons,
					pokemon: pokemons[0],
					showPokemon: false,
					showResult: false,
					correctAnswer: null,
					message: '',
				}
			},
		})

		await wrapper.vm.checkAnswer(1)
		expect(wrapper.find('h2').exists()).toBeTruthy()
		expect(wrapper.vm.showPokemon).toBe(true)

		await wrapper.vm.checkAnswer(2)
		expect(wrapper.vm.message).toBe(`Sorry, it was ${pokemons[0].name}!`)
	})
})
