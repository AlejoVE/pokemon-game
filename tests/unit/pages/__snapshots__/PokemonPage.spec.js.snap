// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Component should match snapshot when pokemon load 1`] = `
<div>
  <h1>Which pokemon is this?</h1>
  <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]" selected="false"></pokemon-options-stub>
</div>
<!--v-if-->
`;
